README.md

![MP3 Voice Stamp Logo](docs/img/banner.png)

 Athletes' companion: adds synthetized voice overlay with various info and on-going timer to your audio files

 [![CodeFactor Code Rating](https://www.codefactor.io/repository/github/MarcinOrlowski/Mp3VoiceStamp/badge?style=flat-square)](https://www.codefactor.io/repository/github/marcinorlowski/mp3voicestamp)
 [![Codacy Badge](https://api.codacy.com/project/badge/Grade/438ca557fec747df965d81bc5958fe7a)](https://www.codacy.com/app/MarcinOrlowski/Mp3VoiceStamp?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=MarcinOrlowski/Mp3VoiceStamp&amp;utm_campaign=Badge_Grade)
 [![Reviewed by Hound](https://img.shields.io/badge/Reviewed_by-Hound-8E64B0.svg)](https://houndci.com)
 
## Table of contents ##

 * [Introduction](#introduction)
 * [Demo MP3 file](#sample)
 * [Features](#features)
 * [Configuration and usage](docs/USAGE.md)
 * [Requirements and installation](docs/INSTALL.md)
 * [Bugs reports and contributions](docs/CONTRIBUTING.md)
 * [Credits and license](#credits-and-license)
 * [What's new?](docs/CHANGES.md)

## Introduction ##
 
 Whenever you do any physical activities, there's high chance you bring your music with you. So do I, but I do swim 
 a lot and in my case, my audio player UI is pretty non existent. It has just 5 control buttons and that's it.
 You cannot even tell what track you selected, unless you recognize it. I personally prefer to swim with 
 so-called DJ mixes - 1 hour long (or more) mixed tracks of various artists just to keep music go without gaps
 between tracks. Having no UI in audio player brings at least two problems:
 
  * it is sometimes hard to tell what mix you just started to listen to (they often need some time to spin up or 
    feature the same opening intro jingle),
  * as there's no display to peek, you cannot easily tell how much time passed so you cannot tell how long you are 
    swimming already.
 
 The aim of this tool is to address these issues. The idea behind is quite simple: you take MP3 track of your choice
 and add synthetized voice overlay with required information. The voice announces the track (by title, filename
 or ID3 tags of your choice), then can keep speaking time passed at given intervals (each 5 minutes by default).

 > ![Tip!](docs/img/tip-small.png)
 > _while swimming is that while title announcement feature is just "useful thing" and nothing more, then the 
 > "time ticks" works more like your personal trainer and for me it adds some motivational bits to my swimming which 
 > is completely unexpected. It seems like it puts kind of subtle pressure so I actually swim more, trying to hear
 > next time "milestone" announced.  It's like some sort of gamification in swimming in practice... Interesting :)_

## Sample ##

 ![Sample MP3](docs/img/music.png) One audio file tells more that thousands words, so why not try the example MP3 file now?
 Hear for yourself what this tool does in practice!

 * [Listen on SoundCloud](https://soundcloud.com/marcinorlowski/mp3voicestamp-tool-demo)
 * [Download sample MP3 file](sample/mp3voicestamp-demo.mp3) (6 MiB size)

 ----
 
 ![OceanPlanet Cover](docs/img/oceanplanet.jpg)
 And if you are interested in music, sample is based on Olga Misty's DJ mix "Ocean Planet 086 Part 2" which can be found on 
 [MixCloud](https://www.mixcloud.com/olgamisty/olga-misty-ocean-planet-086-part-2-aug-06-2018-on-proton-radio/) and
 [SoundCloud](https://soundcloud.com/olga_misty/olga-misty-ocean-planet-086-part-2-aug-06-2018-on-proton-radio).

## Features ##

 * Automatic voice volume adjustment based on music file volume
 * Supports batch processing of multiple files
 * Features dry-run mode for quick tests
 * Highly customizable
 * Minimal runtime dependencies
 * Works on Linux, Mac OS and Windows
 * Free

## Credits and license ##

 * Written and copyrighted &copy;2018 by Marcin Orlowski <mail (#) marcinorlowski (.) com>
 * MP3AudioStamp is open-sourced software licensed under the [MIT license](http://opensource.org/licenses/MIT)
 * Icons used in logo taken from free [Icons8 Olympics Sports Icon Pack](https://icons8.com/free-icons/olympics_sports)



CHANGES.md
dev
---
 * Corrected config file examples
 * MP3 failure returns with non-zero RC when only one input file is provided
 * Improved track title generation from file name

v1.1.0 (2018-10-04)
-------------------
 * Added support for batch processing of multiple audio files
 * Added --verbose and muted most of the messages by default
 * Complete rewrite for better internal architecture
 * Speech speed is now configurable
 * Added --out-format to customize how auto-generated files are named
 * Added --title-format for customized track title announcement format
 * Added support for config files for easier batch processing
 * Changed command line option name 

v1.0.0 (2018-10-01)
-------------------
 * Initial release


mp3voicestamp_app/args.py
# coding=utf8

"""

 MP3 Voice Stamp

 Athletes' companion: adds synthetized voice overlay with various
 info and on-going timer to your audio files

 Copyright ©2018 Marcin Orlowski <mail [@] MarcinOrlowski.com>

 https://github.com/MarcinOrlowski/Mp3VoiceStamp

"""

import argparse
from argparse import RawDescriptionHelpFormatter

from mp3voicestamp_app.config import Config
from mp3voicestamp_app.const import *


class Args(object):
    """Handles command line arguments"""

    @staticmethod
    def parse_args(config):
        """Parses command line arguments

        :returns argsparse
        """
        parser = argparse.ArgumentParser(
            prog=APP_NAME.lower(),
            description='{app} v{v} ({rd})\n'.format(app=APP_NAME, v=VERSION, rd=RELEASE_DATE) +
                        'Adds spoken overlay to MP3 with title, time stamps and more.\n'
                        'Written by Marcin Orlowski <mail@marcinOrlowski.com>\n'
                        'WWW: https://github.com/MarcinOrlowski/Mp3VoiceStamp',
            formatter_class=RawDescriptionHelpFormatter)

        group = parser.add_argument_group('In/Out files')
        group.add_argument(
            '-i', '--in',
            metavar="MP3_FILE", action='store', dest="files_in", nargs='+',
            help="On or more source MP3 files"
        )
        group.add_argument(
            '-o', '--out',
            metavar="DIR/MP3_FILE", action='store', dest="file_out", nargs=1,
            help='Optional output file name or target directory if "-in" option used with multiple files. ' +
                 'If not specified, file name will be generated.'
        )
        group.add_argument(
            '-of', '--out-format', action='store', dest='file_out_format', nargs=1, metavar='FORMAT',
            help='Format string used to generate name of output files. ' +
                 'Default is "{}". '.format(Config.DEFAULT_FILE_OUT_FORMAT) +
                 'See docs for available placeholders.')

        group = parser.add_argument_group('Spoken track title')
        group.add_argument(
            '-tp', '--title-format', action='store', dest='title_format', nargs=1, metavar='FORMAT',
            help='Format string used to generate track title to be spoken. ' +
                 'Default is "{}". '.format(Config.DEFAULT_TITLE_FORMAT) +
                 'See docs for available placeholders.')

        group = parser.add_argument_group('Spoken time ticks')
        # noinspection PyTypeChecker
        group.add_argument(
            '-ti', '--tick-interval', action='store', type=int, dest='tick_interval', nargs=1, metavar='MINUTES',
            help='Interval (in minutes) between spoken ticks. Default is {}.'.format(Config.DEFAULT_TICK_INTERVAL))
        # noinspection PyTypeChecker
        group.add_argument(
            '-to', '--tick-offset', action='store', type=int, dest='tick_offset', nargs=1, metavar='MINUTES',
            help='Offset (in minutes) for first spoken tick. Default is {}.'.format(Config.DEFAULT_TICK_OFFSET))
        group.add_argument(
            '-tf', '--tick-format', action='store', dest='tick_format', nargs=1, metavar='FORMAT',
            help='Format string for spoken time ticks.' +
                 'Default is "{}". '.format(Config.DEFAULT_TICK_FORMAT) +
                 'See docs for available placeholders.')

        group = parser.add_argument_group('Voice synthesizer')
        group.add_argument(
            '-sv', '--speech-volume', action='store', dest='speech_volume_factor', nargs=1, metavar='FLOAT',
            help='Speech volume adjustment multiplier, relative to calculated value. ' +
                 'I.e. "0.5" would lower the volume 50%%, while "2" boost it up to make it twice as loud ' +
                 'as it would be by default. Default is {}.'.format(Config.DEFAULT_SPEECH_VOLUME_FACTOR))
        # noinspection PyTypeChecker
        group.add_argument(
            '-ss', '--speech-speed', action='store', dest='speech_speed', nargs=1, type=int, metavar='INTEGER',
            help='Speech speed in words per minute, in range from {} to {}. Default is {}.'.format(
                Config.SPEECH_SPEED_MIN, Config.SPEECH_SPEED_MAX, Config.DEFAULT_SPEECH_SPEED))

        group = parser.add_argument_group('Configuration')
        group.add_argument(
            '-c', '--config', action='store', dest='config_name', metavar='INI_FILE',
            help='Name of (optional) configuration file to load. If not specified, defaults will be used.'
        )

        group.add_argument(
            '-cs', '--config-save', action='store', dest='config_save_name', metavar='INI_FILE',
            help='Name of configuration file to dump current configuration to.'
        )

        group = parser.add_argument_group('Misc')
        group.add_argument(
            '--dry-run', action='store_true', dest='dry_run_mode',
            help='Simulates processing of the files, printing information on how real files would be processed.'
        )
        group.add_argument(
            '-f', '--force', action='store_true', dest='force',
            help='Forces overwrite of existing output file.'
        )
        group.add_argument(
            '-v', '--verbose', action='store_true', dest='verbose',
            help='Enables verbose output.'
        )
        group.add_argument(
            '--version', action='version', version='{app} v{v} ({rd})'.format(app=APP_NAME, v=VERSION, rd=RELEASE_DATE)
        )

        group = parser.add_argument_group('Developer tools')
        group.add_argument(
            '-d', '--debug', action='store_true', dest='debug',
            help='Enables debug mode.'
        )
        group.add_argument(
            '-nc', '--no-cleanup', action='store_true', dest='no_cleanup',
            help='Do not remove working files and folders on exit.'
        )

        # this trick is to enforce stacktrace in case parse_args() fail (which should normally not happen)
        old_config_debug = config.debug
        if not config.debug:
            config.debug = True

        args = parser.parse_args()

        config.debug = old_config_debug

        if args.files_in is None and args.config_save_name is None:
            parser.print_usage()
            raise ValueError('You must provide at least one MP3 file.')

        config.load(args.config_name)

        config.force_overwrite = args.force
        config.dry_run_mode = args.dry_run_mode
        config.debug = args.debug
        config.no_cleanup = args.no_cleanup
        config.verbose = args.verbose

        config.speech_volume_factor = args.speech_volume_factor
        config.speech_speed = args.speech_speed

        config.tick_interval = args.tick_interval
        config.tick_offset = args.tick_offset
        config.tick_format = args.tick_format

        config.title_format = args.title_format

        # we also support globing (as Windows' cmd is lame as usual)
        config.files_in = []
        # ./mp3vs -i mp3/Olga\ Misty\ -\ Ocean\ Planet\ 086\ Part\ 1\ \[2018-08-06\]\ on\ Proton\ Radio.mp3
        # fails
        # if args.files_in is not None:
        #   import glob
        #   _ = [config.files_in.extend(glob.glob(file_in)) for file_in in args.files_in]
        config.files_in = args.files_in

        config.file_out = args.file_out
        config.file_out_format = args.file_out_format

        return args
